go/signatures_test.go

```diff
@@ -0,0 +1,336 @@
+}
+
+func sigTestCertLogEntry(t *testing.T) LogEntry {
+	return LogEntry{
+		Index: 0,
+		Leaf: MerkleTreeLeaf{
+			Version:  V1,
+			LeafType: TimestampedEntryLeafType,
+			TimestampedEntry: TimestampedEntry{
+				Timestamp: sigTestSCTTimestamp,
+				EntryType: X509LogEntryType,
+				X509Entry: mustDehex(t, sigTestDERCertString),
+			},
+		},
+	}
+}
+
+func mustCreateSignatureVerifier(t *testing.T, pk crypto.PublicKey) SignatureVerifier {
+	sv, err := NewSignatureVerifier(pk)
+	if err != nil {
+		t.Fatalf("Failed to create SignatureVerifier: %v", err)
+	}
+	return *sv
+}
+
+func TestVerifySCTSignatureEC(t *testing.T) {
+	v := mustCreateSignatureVerifier(t, sigTestECPublicKey(t))
+	if err := v.VerifySCTSignature(sigTestSCTEC(t), sigTestCertLogEntry(t)); err != nil {
+		t.Fatalf("Failed to verify signature on SCT: %v", err)
+	}
+
+}
+
+func TestVerifySCTSignatureRSA(t *testing.T) {
+	v := mustCreateSignatureVerifier(t, sigTestRSAPublicKey(t))
+	if err := v.VerifySCTSignature(sigTestSCTRSA(t), sigTestCertLogEntry(t)); err != nil {
+		t.Fatalf("Failed to verify signature on SCT: %v", err)
+	}
+
+}
+
+func TestVerifySCTSignatureFailsForMismatchedSignatureAlgorithm(t *testing.T) {
+	v := mustCreateSignatureVerifier(t, sigTestECPublicKey(t))
+	if err := v.VerifySCTSignature(sigTestSCTRSA(t), sigTestCertLogEntry(t)); err == nil {
+		t.Fatalf("Sucessfully verified with mismatched signature algorithm")
+	}
+
+}
+
+func TestVerifySCTSignatureFailsForUnknownSignatureAlgorithm(t *testing.T) {
+	v := mustCreateSignatureVerifier(t, sigTestECPublicKey(t))
+	if err := v.VerifySCTSignature(sigTestSCTWithSignature(t, sigTestCertSCTSignatureUnsupportedSignatureAlgorithm, sigTestKeyIDEC), sigTestCertLogEntry(t)); err == nil {
+		t.Fatal("Successfully verified signature with unsupported signature algorithm")
+	}
+}
+
+func TestVerifySCTSignatureFailsForUnknownHashAlgorithm(t *testing.T) {
+	v := mustCreateSignatureVerifier(t, sigTestECPublicKey(t))
+	if err := v.VerifySCTSignature(sigTestSCTWithSignature(t, sigTestCertSCTSignatureUnsupportedHashAlgorithm, sigTestKeyIDEC), sigTestCertLogEntry(t)); err == nil {
+		t.Fatal("Successfully verified signature with unsupported hash algorithm")
+	}
+}
+
+func testVerifySCTSignatureFailsForIncorrectInput(t *testing.T, sct SignedCertificateTimestamp, sv SignatureVerifier) {
+	entry := sigTestCertLogEntry(t)
+	for i := range entry.Leaf.TimestampedEntry.X509Entry {
+		old := entry.Leaf.TimestampedEntry.X509Entry[i]
+		entry.Leaf.TimestampedEntry.X509Entry[i] ^= byte(mrand.Intn(255) + 1)
+		if err := sv.VerifySCTSignature(sct, entry); err == nil {
+			t.Fatalf("Incorrectly verfied signature over corrupted input data, uncovered byte at %d?", i)
+		}
+		entry.Leaf.TimestampedEntry.X509Entry[i] = old
+	}
+}
+
+func testVerifySCTSignatureFailsForIncorrectSignature(t *testing.T, sct SignedCertificateTimestamp, sv SignatureVerifier) {
+	sct.Signature.Signature[mrand.Intn(len(sct.Signature.Signature))] ^= byte(mrand.Intn(255) + 1)
```

Comment:
To echo Ben's comment, why not change a particular byte in a predictable way? What does the randomness here buy?